// @fremtind/jkl-card-react/build/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var jklButtonReact = require('@fremtind/jkl-button-react');

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var classnames = createCommonjsModule(function (module) {
/*!
  Copyright (c) 2017 Jed Watson.
  Licensed under the MIT License (MIT), see
  http://jedwatson.github.io/classnames
*/
/* global define */

(function () {

	var hasOwn = {}.hasOwnProperty;

	function classNames () {
		var classes = [];

		for (var i = 0; i < arguments.length; i++) {
			var arg = arguments[i];
			if (!arg) continue;

			var argType = typeof arg;

			if (argType === 'string' || argType === 'number') {
				classes.push(arg);
			} else if (Array.isArray(arg) && arg.length) {
				var inner = classNames.apply(null, arg);
				if (inner) {
					classes.push(inner);
				}
			} else if (argType === 'object') {
				for (var key in arg) {
					if (hasOwn.call(arg, key) && arg[key]) {
						classes.push(key);
					}
				}
			}
		}

		return classes.join(' ');
	}

	if ( module.exports) {
		classNames.default = classNames;
		module.exports = classNames;
	} else {
		window.classNames = classNames;
	}
}());
});
var classnames_1 = classnames.named;

var Card = function Card(_ref) {
  var title = _ref.title,
      children = _ref.children,
      className = _ref.className,
      media = _ref.media,
      action = _ref.action,
      dark = _ref.dark,
      clickable = _ref.clickable;
  var componentClassName = classnames("jkl-card", className, {
    "jkl-card--dark": dark,
    "jkl-card--clickable": clickable
  });
  var Button = jklButtonReact.PrimaryButton;

  if (action) {
    switch (action.type) {
      case "secondary":
        Button = jklButtonReact.SecondaryButton;
        break;

      case "tertiary":
        Button = jklButtonReact.TertiaryButton;
        break;
    }
  }

  return /*#__PURE__*/React.createElement("div", {
    "data-testid": "jkl-card",
    className: componentClassName
  }, media && /*#__PURE__*/React.createElement("img", {
    className: "jkl-card__media",
    src: media.src,
    alt: media.alt
  }), title && /*#__PURE__*/React.createElement("div", {
    "data-testid": "jkl-card__title",
    className: "jkl-card__title jkl-h3"
  }, clickable ? /*#__PURE__*/React.createElement("a", {
    className: "jkl-card__link",
    href: clickable.href,
    onClick: clickable.onClick
  }, title) : title), /*#__PURE__*/React.createElement("div", {
    className: "jkl-card__children"
  }, children), action && /*#__PURE__*/React.createElement("div", {
    className: "jkl-card__action"
  }, /*#__PURE__*/React.createElement(Button, {
    onClick: action.onClick
  }, action.name)));
};

exports.Card = Card;
