// @fremtind/jkl-accordion-react/build/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var CoreToggle = _interopDefault(require('@nrk/core-toggle/jsx'));
var React = require('react');
var React__default = _interopDefault(React);
var jklReactHooks = require('@fremtind/jkl-react-hooks');

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return;
  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var classnames = createCommonjsModule(function (module) {
/*!
  Copyright (c) 2017 Jed Watson.
  Licensed under the MIT License (MIT), see
  http://jedwatson.github.io/classnames
*/
/* global define */

(function () {

	var hasOwn = {}.hasOwnProperty;

	function classNames () {
		var classes = [];

		for (var i = 0; i < arguments.length; i++) {
			var arg = arguments[i];
			if (!arg) continue;

			var argType = typeof arg;

			if (argType === 'string' || argType === 'number') {
				classes.push(arg);
			} else if (Array.isArray(arg) && arg.length) {
				var inner = classNames.apply(null, arg);
				if (inner) {
					classes.push(inner);
				}
			} else if (argType === 'object') {
				for (var key in arg) {
					if (hasOwn.call(arg, key) && arg[key]) {
						classes.push(key);
					}
				}
			}
		}

		return classes.join(' ');
	}

	if ( module.exports) {
		classNames.default = classNames;
		module.exports = classNames;
	} else {
		window.classNames = classNames;
	}
}());
});
var classnames_1 = classnames.named;

function AccordionItem(_ref) {
  var children = _ref.children,
      title = _ref.title,
      className = _ref.className,
      _ref$startExpanded = _ref.startExpanded,
      startExpanded = _ref$startExpanded === void 0 ? false : _ref$startExpanded;

  var _useState = React.useState(startExpanded),
      _useState2 = _slicedToArray(_useState, 2),
      isOpen = _useState2[0],
      setIsOpen = _useState2[1];

  var _useAnimatedHeight = jklReactHooks.useAnimatedHeight(isOpen),
      _useAnimatedHeight2 = _slicedToArray(_useAnimatedHeight, 1),
      elementRef = _useAnimatedHeight2[0];

  var componentClassName = classnames("jkl-accordion-item", className, {
    "jkl-accordion-item--expanded": isOpen
  });

  var onToggle = function onToggle() {
    return setIsOpen(!isOpen);
  };

  return /*#__PURE__*/React__default.createElement("div", {
    "data-testid": "jkl-accordion-item",
    className: componentClassName
  }, /*#__PURE__*/React__default.createElement("button", {
    className: "jkl-accordion-item__title",
    type: "button"
  }, /*#__PURE__*/React__default.createElement("span", {
    className: "jkl-accordion-item__title-text"
  }, title), /*#__PURE__*/React__default.createElement("span", {
    className: "jkl-accordion-item__title-icon"
  })), /*#__PURE__*/React__default.createElement(CoreToggle, {
    ref: elementRef,
    "data-testid": "jkl-accordion-item__content-wrapper",
    className: "jkl-accordion-item__content-wrapper",
    hidden: !isOpen,
    onToggle: onToggle
  }, /*#__PURE__*/React__default.createElement("div", {
    className: "jkl-accordion-item__content"
  }, children)));
}

function Accordion(_ref) {
  var children = _ref.children,
      _ref$className = _ref.className,
      className = _ref$className === void 0 ? "" : _ref$className;
  return /*#__PURE__*/React__default.createElement("div", {
    "data-testid": "jkl-accordion",
    className: "jkl-accordion ".concat(className)
  }, children);
}

exports.Accordion = Accordion;
exports.AccordionItem = AccordionItem;
